# Strava Fitness & Wellness Analysis

### Data Analysis Project

This project analyzes user activity, sleep, and weight/BMI data to understand fitness behavior, engagement patterns, and activity trends.

## Business Problem

A fitness and wellness company collects data from users through fitness tracking devices. The data includes daily activity, sleep patterns, and weight/BMI information.

The company wants to understand how users engage with fitness tracking and identify patterns in their activity and wellness behaviour.

The analysis will help answer questions such as:

- How active are users on a typical day?
- How consistent is user activity?
- How do activity levels vary across days?
- What are the users' sleep patterns?
- What weight and BMI patterns can be observed among users who track their weight?
- What insights can help the company improve user engagement and encourage healthier, more consistent behaviour?

## Project Objectives

The main objectives of this analysis are to:

1. Analyze users' daily activity levels using steps, distance, calories, and activity minutes.
2. Identify patterns in user activity across different days of the week and over the tracking period.
3. Analyze sleep duration and time spent in bed.
4. Examine weight and BMI patterns among users who recorded weight data.
5. Identify differences in activity levels across users.
6. Generate data-driven insights that can help improve user engagement and support healthier, more consistent fitness behaviour.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df_activity=pd.read_csv("dailyActivity_merged.csv")
df_sleep=pd.read_csv("sleepDay_merged.csv")
df_weight=pd.read_csv("weightLogInfo_merged.csv")

In [6]:
df_activity.shape

(940, 15)

In [7]:
df_sleep.shape

(413, 5)

In [8]:
df_weight.shape

(67, 8)

In [9]:
df_activity.columns

Index(['Id', 'ActivityDate', 'TotalSteps', 'TotalDistance', 'TrackerDistance',
       'LoggedActivitiesDistance', 'VeryActiveDistance',
       'ModeratelyActiveDistance', 'LightActiveDistance',
       'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes',
       'LightlyActiveMinutes', 'SedentaryMinutes', 'Calories'],
      dtype='object')

In [10]:
df_sleep.columns

Index(['Id', 'SleepDay', 'TotalSleepRecords', 'TotalMinutesAsleep',
       'TotalTimeInBed'],
      dtype='object')

In [11]:
df_weight.columns

Index(['Id', 'Date', 'WeightKg', 'WeightPounds', 'Fat', 'BMI',
       'IsManualReport', 'LogId'],
      dtype='object')

In [12]:
df_activity.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,4/12/2016,13162,8.50,8.50,0.0,1.88,0.55,6.06,0.0,25,13,328,728,1985
1,1503960366,4/13/2016,10735,6.97,6.97,0.0,1.57,0.69,4.71,0.0,21,19,217,776,1797
2,1503960366,4/14/2016,10460,6.74,6.74,0.0,2.44,0.40,3.91,0.0,30,11,181,1218,1776
3,1503960366,4/15/2016,9762,6.28,6.28,0.0,2.14,1.26,2.83,0.0,29,34,209,726,1745
4,1503960366,4/16/2016,12669,8.16,8.16,0.0,2.71,0.41,5.04,0.0,36,10,221,773,1863


In [13]:
df_sleep.head()

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
0,1503960366,4/12/2016 12:00:00 AM,1,327,346
1,1503960366,4/13/2016 12:00:00 AM,2,384,407
2,1503960366,4/15/2016 12:00:00 AM,1,412,442
3,1503960366,4/16/2016 12:00:00 AM,2,340,367
4,1503960366,4/17/2016 12:00:00 AM,1,700,712


In [14]:
df_weight.head()

,Id,Date,WeightKg,WeightPounds,Fat,BMI,IsManualReport,LogId
0,1503960366,5/2/2016 11:59:59 PM,52.599998,115.963147,22.0,22.650000,True,1462233599000
1,1503960366,5/3/2016 11:59:59 PM,52.599998,115.963147,NaN,22.650000,True,1462319999000
2,1927972279,4/13/2016 1:08:52 AM,133.500000,294.317120,NaN,47.540001,False,1460509732000
3,2873212765,4/21/2016 11:59:59 PM,56.700001,125.002104,NaN,21.450001,True,1461283199000
4,2873212765,5/12/2016 11:59:59 PM,57.299999,126.324875,NaN,21.690001,True,1463097599000


In [55]:
print("Activity Dataset Info:")
df_activity.info()

print("\nSleep Dataset Info:")
df_sleep.info()

print("\nWeight Dataset Info:")
df_weight.info()

Activity Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Id                        940 non-null    int64         
 1   ActivityDate              940 non-null    datetime64[ns]
 2   TotalSteps                940 non-null    int64         
 3   TotalDistance             940 non-null    float64       
 4   TrackerDistance           940 non-null    float64       
 5   LoggedActivitiesDistance  940 non-null    float64       
 6   VeryActiveDistance        940 non-null    float64       
 7   ModeratelyActiveDistance  940 non-null    float64       
 8   LightActiveDistance       940 non-null    float64       
 9   SedentaryActiveDistance   940 non-null    float64       
 10  VeryActiveMinutes         940 non-null    int64         
 11  FairlyActiveMinutes       940 non-null    int64         
 12 

## Data Structure and Quality Check

The datasets were inspected to understand their structure, data types, and completeness.

- The Activity dataset contains 940 records and 15 columns with no missing values.
- The Sleep dataset initially contains 413 records and 5 columns with no missing values.
- The Weight dataset contains 67 records and 8 columns.
- The `Fat` column in the Weight dataset has only 2 available values out of 67 records, making it unsuitable for meaningful analysis.
- The date columns are initially stored as object/string data types and will be standardized before further analysis.

In [20]:
df_activity["ActivityDate"]=pd.to_datetime(df_activity["ActivityDate"]).dt.date

In [23]:
df_sleep["SleepDay"]=pd.to_datetime(df_sleep["SleepDay"]).dt.date

In [25]:
df_weight["Date"]=pd.to_datetime(df_weight["Date"]).dt.date

## Date Standardization

The date columns were converted into a consistent date format for analysis.

- `ActivityDate` was converted to date format.
- `SleepDay` was converted to date format.
- `Date` in the Weight dataset was converted to date format.
- The time component was removed because the analysis focuses on daily-level activity, sleep, and weight records.

This standardization also prepares the datasets for date-based analysis and future SQL joins.

In [29]:
df_activity.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,2016-04-12,13162,8.50,8.50,0.0,1.88,0.55,6.06,0.0,25,13,328,728,1985
1,1503960366,2016-04-13,10735,6.97,6.97,0.0,1.57,0.69,4.71,0.0,21,19,217,776,1797
2,1503960366,2016-04-14,10460,6.74,6.74,0.0,2.44,0.40,3.91,0.0,30,11,181,1218,1776
3,1503960366,2016-04-15,9762,6.28,6.28,0.0,2.14,1.26,2.83,0.0,29,34,209,726,1745
4,1503960366,2016-04-16,12669,8.16,8.16,0.0,2.71,0.41,5.04,0.0,36,10,221,773,1863


In [30]:
df_sleep.head()

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
0,1503960366,2016-04-12,1,327,346
1,1503960366,2016-04-13,2,384,407
2,1503960366,2016-04-15,1,412,442
3,1503960366,2016-04-16,2,340,367
4,1503960366,2016-04-17,1,700,712


In [31]:
df_weight.head()

,Id,Date,WeightKg,WeightPounds,Fat,BMI,IsManualReport,LogId
0,1503960366,2016-05-02,52.599998,115.963147,22.0,22.650000,True,1462233599000
1,1503960366,2016-05-03,52.599998,115.963147,NaN,22.650000,True,1462319999000
2,1927972279,2016-04-13,133.500000,294.317120,NaN,47.540001,False,1460509732000
3,2873212765,2016-04-21,56.700001,125.002104,NaN,21.450001,True,1461283199000
4,2873212765,2016-05-12,57.299999,126.324875,NaN,21.690001,True,1463097599000


In [34]:
df_activity.duplicated().sum()

0

In [35]:
df_sleep.duplicated().sum()

3

In [36]:
df_weight.duplicated().sum()

0

In [37]:
df_sleep = df_sleep.drop_duplicates()

In [38]:
df_sleep.duplicated().sum()

0

In [39]:
df_activity['Id'].nunique()

33

In [40]:
df_sleep['Id'].nunique()

24

In [41]:
df_weight['Id'].nunique()

8

## User Coverage

The number of unique users was examined across each dataset to understand data coverage.

- Activity dataset: 33 unique users
- Sleep dataset: 24 unique users
- Weight dataset: 8 unique users

The datasets do not contain the same number of users. Therefore, cross-dataset analysis should consider the available user coverage when interpreting results.

In [42]:
df_activity[['TotalSteps', 'TotalDistance', 'Calories']].describe()

,TotalSteps,TotalDistance,Calories
count,940.000000,940.000000,940.000000
mean,7637.910638,5.489702,2303.609574
std,5087.150742,3.924606,718.166862
min,0.000000,0.000000,0.000000
25%,3789.750000,2.620000,1828.500000
50%,7405.500000,5.245000,2134.000000
75%,10727.000000,7.712500,2793.250000
max,36019.000000,28.030001,4900.000000


### Overall Activity Summary

- The dataset contains 940 daily activity records.
- Average daily steps are approximately 7,638.
- Average daily distance is approximately 5.49 km.
- Average daily calories burned are approximately 2,304 kcal.
- Daily steps vary considerably, ranging from 0 to 36,019.
- The median daily step count is approximately 7,406.

In [43]:
df_activity[['VeryActiveMinutes',
             'FairlyActiveMinutes',
             'LightlyActiveMinutes',
             'SedentaryMinutes']].mean()

VeryActiveMinutes        21.164894
FairlyActiveMinutes      13.564894
LightlyActiveMinutes    192.812766
SedentaryMinutes        991.210638
dtype: float64

### Activity Level Summary

- Users recorded an average of 21.2 very active minutes per day.
- Users recorded an average of 13.6 fairly active minutes per day.
- Users recorded an average of 192.8 lightly active minutes per day.
- Sedentary time was the largest activity category, averaging 991.2 minutes per day.

## Sleep Analysis

The Sleep dataset was analyzed to understand users' sleep duration and time spent in bed.

The analysis focuses on average sleep duration and the difference between time spent asleep and total time in bed.

In [44]:
df_sleep[['TotalMinutesAsleep', 'TotalTimeInBed']].describe()

,TotalMinutesAsleep,TotalTimeInBed
count,410.000000,410.000000
mean,419.173171,458.482927
std,118.635918,127.455140
min,58.000000,61.000000
25%,361.000000,403.750000
50%,432.500000,463.000000
75%,490.000000,526.000000
max,796.000000,961.000000


### Sleep Summary

- The dataset contains 410 sleep records after removing duplicate records.
- Average sleep duration is approximately 7.0 hours per day.
- Median sleep duration is approximately 7.2 hours per day.
- Users spent an average of approximately 7.6 hours in bed per day.

In [45]:
df_activity[['TotalSteps', 'TotalDistance', 'Calories',
             'VeryActiveMinutes', 'FairlyActiveMinutes',
             'LightlyActiveMinutes', 'SedentaryMinutes']].mean()

TotalSteps              7637.910638
TotalDistance              5.489702
Calories                2303.609574
VeryActiveMinutes         21.164894
FairlyActiveMinutes       13.564894
LightlyActiveMinutes     192.812766
SedentaryMinutes         991.210638
dtype: float64

In [46]:
df_activity['TotalSteps'].describe()

count      940.000000
mean      7637.910638
std       5087.150742
min          0.000000
25%       3789.750000
50%       7405.500000
75%      10727.000000
max      36019.000000
Name: TotalSteps, dtype: float64

In [51]:
print(df_activity[['Id', 'ActivityDate']].head())
print(df_sleep[['Id', 'SleepDay']].head())

           Id ActivityDate
0  1503960366   2016-04-12
1  1503960366   2016-04-13
2  1503960366   2016-04-14
3  1503960366   2016-04-15
4  1503960366   2016-04-16
           Id    SleepDay
0  1503960366  2016-04-12
1  1503960366  2016-04-13
2  1503960366  2016-04-15
3  1503960366  2016-04-16
4  1503960366  2016-04-17


In [47]:
active_minutes = (
    df_activity['VeryActiveMinutes']
    + df_activity['FairlyActiveMinutes']
    + df_activity['LightlyActiveMinutes']
)

print("Average active minutes:", active_minutes.mean())
print("Average sedentary minutes:", df_activity['SedentaryMinutes'].mean())

Average active minutes: 227.54255319148936
Average sedentary minutes: 991.2106382978724


In [48]:
user_steps = (
    df_activity.groupby('Id')['TotalSteps']
    .mean()
    .sort_values(ascending=False)
)

print("Most active users:")
print(user_steps.head())

print("\nLeast active users:")
print(user_steps.tail())

Most active users:
Id
8877689391    16040.032258
8053475328    14763.290323
1503960366    12116.741935
2022484408    11370.645161
7007744171    11323.423077
Name: TotalSteps, dtype: float64

Least active users:
Id
1844505072    2580.064516
6775888955    2519.692308
4020332650    2267.225806
8792009665    1853.724138
1927972279     916.129032
Name: TotalSteps, dtype: float64


## User Activity Variation

Average daily steps were calculated for each user to understand differences in activity levels.

This analysis helps identify variation in activity behaviour across users and highlights differences between users with higher and lower average step counts.

In [49]:
df_activity['ActivityDate'] = pd.to_datetime(df_activity['ActivityDate'])

df_activity['DayOfWeek'] = df_activity['ActivityDate'].dt.day_name()

df_activity.groupby('DayOfWeek')['TotalSteps'].mean()

DayOfWeek
Friday       7448.230159
Monday       7780.866667
Saturday     8152.975806
Sunday       6933.231405
Thursday     7405.836735
Tuesday      8125.006579
Wednesday    7559.373333
Name: TotalSteps, dtype: float64

## Activity by Day of Week

Average daily steps were compared across the days of the week to identify differences in activity levels throughout the week.

This helps understand whether users tend to be more or less active on particular days.

In [52]:
df_weight['BMI'].describe()

count    67.000000
mean     25.185224
std       3.066963
min      21.450001
25%      23.959999
50%      24.389999
75%      25.559999
max      47.540001
Name: BMI, dtype: float64

In [53]:
df_weight['WeightKg'].describe()

count     67.000000
mean      72.035821
std       13.923206
min       52.599998
25%       61.400002
50%       62.500000
75%       85.049999
max      133.500000
Name: WeightKg, dtype: float64

## Weight and BMI Analysis

The Weight dataset was analyzed to understand the weight and BMI characteristics of users who recorded their weight data.

BMI and weight distributions were examined using descriptive statistics.

In [57]:
df_activity.drop(columns=["DayOfWeek"]).to_csv(
    "activity_cleaned.csv",
    index=False
)

In [58]:
df_sleep.to_csv("sleep_cleaned.csv", index=False)

In [59]:
df_weight.to_csv("weight_cleaned.csv",index=False)